zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -0,0 +1,295 @@
+
+  BackupStorageProvider storage;
+  FileTxnSnapLog snapLog;
+  long zxidToRestore;
+  boolean dryRun;
+
+  List<BackupFileInfo> logs;
+  List<BackupFileInfo> snaps;
+  List<BackupFileInfo> filesToCopy;
+
+  int mostRecentLogNeededIndex;
+  int snapNeededIndex;
+  int oldestLogNeededIndex;
+
+  /**
+   * Default constructor; requires using parseArgs to setup state.
+   */
+  public RestoreFromBackupTool() {
+    this(null, null, -1L, false);
+  }
+
+  /**
+   * Constructor
+   * @param storageProvider the backup provider from which to restore the backups
+   * @param snapLog the snap and log provider to use
+   * @param zxidToRestore the zxid upto which to restore, or Long.MAX to restore to the latest
+   *                      available transactionally consistent zxid.
+   * @param dryRun whether this is a dryrun in which case no files are actually copied
+   */
+  RestoreFromBackupTool(BackupStorageProvider storageProvider, FileTxnSnapLog snapLog,
+      long zxidToRestore, boolean dryRun) {
+
+    filesToCopy = new ArrayList<BackupFileInfo>();
+    snapNeededIndex = -1;
+    mostRecentLogNeededIndex = -1;
+    oldestLogNeededIndex = -1;
+
+    this.storage = storageProvider;
+    this.zxidToRestore = zxidToRestore;
+    this.snapLog = snapLog;
+    this.dryRun = dryRun;
+  }
+
+  /**
+   * Attempts to perform a restore.
+   */
+  public void run() throws IOException {
+    if (!findFilesToRestore()) {
+      throw new RestoreException("Failed to find a valid snapshot and logs to restore.");
+    }
+
+    // Create a temporary dir in backup storage for backup file processing
+    File tempDirForRestoration =
+        new File(filesToCopy.listIterator().next().getBackedUpFile().getParent(),
+            "Restore_" + Long.toHexString(zxidToRestore));
+
+    try {
+      storage.processBackupFilesForRestoration(tempDirForRestoration, filesToCopy, zxidToRestore);
+      LOG.info("Copied " + filesToCopy.size() + " backup files to temp dir " + tempDirForRestoration
+          .getPath() + " in backup storage for file preparation for restoration to zxid " + zxidToRestore + ".");
+
+      List<File> processedFiles = BackupStorageUtil.readFilesRecursivelyInDirectory(tempDirForRestoration);
+      for (File processedFile : processedFiles) {
+        String fileName = processedFile.getName();
+        File parentFile = processedFile.getParentFile();
+        File base = Util.isSnapshotFileName(fileName) ? snapLog.getSnapDir() : snapLog.getDataDir();
+        File destination = new File(base, fileName);
+
+        if (!destination.exists()) {
+          LOG.info("Copying " + fileName + " to " + destination.getPath() + ".");
+
+          if (!dryRun) {
+            // Adding "RESTORE_" prefix to file name to differentiate the restore files from original backup files,
+            // so in storage side it will use a different way to construct the file path from original backup files
```

Comment:
This sounds like as if you are adding the prefix to the files you're copying to (files created in local storage as a result of copying). I don't believe this is what you really meant - did you mean that the source files (already) _have_ the RESTORE prefix prepended to differentiate in the _backup (not local) storage_?

Is this differentiation necessary though? We are already differentiating by putting them in a temp directory so we don't mix them up with the original backup files, right?

Perhaps another thing I might want to do here is to add/modify prefixes to differentiate files before/after processing?

Reply:
Discussed offline, this comment along with the next a few comments will be addressed by creating the temp dir in local instead of in backup storage. Creating temp dir in backup storage requires one implementation for each backup storage type (GPFS, HDFS, etc.) and will need special handlings on the files, which is a heavy-weighted operation. Therefore we decided to creating the temp dir in local instead which one implementation can be used for all the backup storages, and does not require special handling like this.